lib/Runtime/Base/ThreadContext.cpp

```diff
@@ -3551,21 +3598,110 @@ ThreadContext::InvalidatePropertyGuardEntry(const Js::PropertyRecord* propertyRe
+    {
+        Js::PropertyGuard* guard = entry->sharedGuard;
+        if (guard->GetValue() == reinterpret_cast<intptr_t>(type))
+        {
+            if (PHASE_TRACE1(Js::TracePropertyGuardsPhase) || PHASE_VERBOSE_TRACE1(Js::FixedMethodsPhase))
+            {
+                Output::Print(_u("FixedFields: invalidating guard: name: %s, address: 0x%p, value: 0x%p, value address: 0x%p\n"),
+                              propertyRecord->GetBuffer(), guard, guard->GetValue(), guard->GetAddressOfValue());
+                Output::Flush();
+            }
+
+            if (PHASE_TESTTRACE1(Js::TracePropertyGuardsPhase) || PHASE_VERBOSE_TESTTRACE1(Js::FixedMethodsPhase))
+            {
+                Output::Print(_u("FixedFields: invalidating guard: name: %s, value: 0x%p\n"), propertyRecord->GetBuffer(), guard->GetValue());
+                Output::Flush();
+            }
+
+            guard->Invalidate();
+        }
+    }
+
+    uint count = 0;
+    entry->uniqueGuards.MapAndRemoveIf([&count, propertyRecord, type](RecyclerWeakReference<Js::PropertyGuard>* guardWeakRef) -> bool
+    {
+        Js::PropertyGuard* guard = guardWeakRef->Get();
+        if (guard != nullptr && guard->GetValue() == reinterpret_cast<intptr_t>(type))
         {
-            // TODO: In future, we should compact the uniqueGuards dictionary so this function can be called from PreCollectionCallback
-            // instead
-            this->ClearInvalidatedUniqueGuards();
-            this->recyclableData->constructorCacheInvalidationCount = 0;
+            if (PHASE_TRACE1(Js::TracePropertyGuardsPhase) || PHASE_VERBOSE_TRACE1(Js::FixedMethodsPhase))
+            {
+                Output::Print(_u("FixedFields: invalidating guard: name: %s, address: 0x%p, value: 0x%p, value address: 0x%p\n"),
+                    propertyRecord->GetBuffer(), guard, guard->GetValue(), guard->GetAddressOfValue());
+                Output::Flush();
+            }
+
+            if (PHASE_TESTTRACE1(Js::TracePropertyGuardsPhase) || PHASE_VERBOSE_TESTTRACE1(Js::FixedMethodsPhase))
+            {
+                Output::Print(_u("FixedFields: invalidating guard: name: %s, value: 0x%p\n"),
+                    propertyRecord->GetBuffer(), guard->GetValue());
+                Output::Flush();
+            }
+
+            guard->Invalidate();
+            count++;
+            return true;
+        }
+        else
+        {
+            return false;
         }
+    });
+
+    if (!isAllPropertyGuardsInvalidation)
+    {
+        this->UpdateConstructorCacheInvalidationCount(count);
     }
 
+    this->UpdateEntryPoints(propertyRecord, entry);
+}
+
+void ThreadContext::UpdateConstructorCacheInvalidationCount(uint count)
+{
+    // Count no. of invalidations done so far. Exclude if this is all property guards invalidation in which case
+    // the unique Guards will be cleared anyway.
+    this->recyclableData->constructorCacheInvalidationCount += count;
```

Comment:
The second copy of this comment seems less relevant, since the logic to "Exclude if..." is up at the call site.

Reply:
Yeah, I'm not sure what the right thing to do here is.  On the one hand, this sentence (which, btw, I inherited from master; I just moved it to a different part of the file) is an important part of the contract of UpdateConstructorCacheInvalidationCount.  On the other, removing this part of the contract and having this method do the check would arguably be a cleaner interface design, but at the cost of adding an additional parameter.  (Plus, it's kind of weird to have a method take a "don't do anything if this value is true" argument.)